src/borg/testsuite/repository.py

```diff
@@ -15,11 +15,29 @@
 from ..remote import RemoteRepository, InvalidRPCMethod, PathNotAllowed
 from ..repository import Repository, LoggedIO, MAGIC, MAX_DATA_SIZE, TAG_DELETE, TAG_PUT2, TAG_PUT, TAG_COMMIT
 from ..repoobj import RepoObj
-from . import BaseTestCase
 from .hashindex import H
 
 
-UNSPECIFIED = object()  # for default values where we can't use None
+@pytest.fixture()
+def repository(tmpdir):
+    # setup
+    repository_location = os.path.join(str(tmpdir), "repository")
+    repository = Repository(repository_location, exclusive=True, create=True)
+    yield repository.__enter__()
+
+    # teardown
+    repository.__exit__(None, None, None)
+    # shutil.rmtree(str(tmpdir))
+
+
+@contextmanager
+def reopen(repository, exclusive=True):
+    try:
+        repository.__exit__(None, None, None)
+        repo = Repository(repository.path, exclusive=exclusive, create=False)
+        yield repo.__enter__()
+    finally:
+        repository.__exit__(None, None, None)
```

Comment:
try try/finally does not look quite right.

guess it should be:
```
r = repo.__enter__()
try:
    yield r
finally:
    repo.__exit__(...)
```

Reply:
also, i think the repository fixture would not work correctly with the reopen CM:
- the fixture would enter the context1 before the test function starts
- reopen would exit context1, enter context2, exit context2
- the fixture would exit context1 after the test function finished

So there is one exit too much and also the stuff is not supposed to be nested into each other - we are dealing with the same repository for both contexts.

Reply:
after thinking a bit about it: if we want to have something like reopen, i don't think we can have an **"entered repository context" as a fixture**.

guess we could have **"repository" as a fixture** though. this still removes some code duplication, could clean up afterwards, etc. and would avoid the issues due to bad nesting. there could be also **"remote_repository"** as a fixture.

Reply:
Yes, that's what I was thinking as well, when discussing this with Ted. Though I don't know the code as well, admittedly. So not sure it's possible speed-wise. But logically it would be best.

Reply:
> So there is one exit too much and also the stuff is not supposed to be nested into each other

This is a really good point, and something I had overlooked. This may be the source of some of the warnings I was encountering.

> after thinking a bit about it: if we want to have something like reopen, i don't think we can have an "entered repository context" as a fixture.

Are you suggesting to remove the `reopen` CM in place of something different (like just a function that reopens and returns?) to avoid poor nesting, or remove `reopen` entirely? 

> guess we could have "repository" as a fixture though. this still removes some code duplication, could clean up afterwards, etc. and would avoid the issues due to bad nesting

To be clear, are you suggesting to have `repository` do the initial setup, yield the repo, at end do final teardown, meanwhile entering the repo within the test instead of the in the repository fixture?

Reply:
the CM would be entered/exited within the test function (not within the fixture code).

the whole point of `reopen()` was to get a really completely fresh Repository instance, so we do not rely on previous state except the files that were stored on disk.